---
title: "Using Phyloseq"
subtitle: "12_phyloseq"
author: "Bonnie Hurwitz"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
```

```{bash, engine.opts='-l'}
# Getting started, Setting the variables we need
NETID="YOUR_NETID"
XFILE="YOUR_XFILE"
```

```{bash, engine.opts='-l'}
# Let's go to the directory where we created our our kraken2/bracken results file
cd /xdisk/bhurwitz/bh_class/${NETID}/assignments/11_taxonomy
```

```{bash, engine.opts='-l'}
# Let's create all of the output directories we need for our analysis
for dir in out_megahit_taxonomy out_metaspades_taxonomy out_reads_taxonomy; do
    mkdir ../12_phyloseq/$dir
done
```

```{bash, engine.opts='-l'}
# Now we can loop through all of the samples and create biom files (for each sample, each data type, and combined across data types)
# for each of our *.result files, for annotation from each assembly and the reads
XFILE_PATH="/xdisk/bhurwitz/bh_class/${NETID}/assignments/05_getting_data/${XFILE}"
for sample in `cat $XFILE_PATH`; do
    ALL=""
    for dir in out_megahit_taxonomy out_metaspades_taxonomy out_reads_taxonomy; do
        ALL=${ALL}" ./$dir/$sample/bracken_results.txt"
        mkdir ../12_phyloseq/$dir/$sample
        /contrib/singularity/shared/bhurwitz/kraken-biom:1.2.0--pyh5e36f6f_0.sif kraken-biom \
        ./$dir/$sample/bracken_results.txt --fmt json -o ../12_phyloseq/$dir/$sample/bracken_results.biom
    done
    /contrib/singularity/shared/bhurwitz/kraken-biom:1.2.0--pyh5e36f6f_0.sif kraken-biom \
    ${ALL} --fmt json -o ../12_phyloseq/$sample.biom
    # copy the last biom file to use in exploring phyloseq below.
    cp ../12_phyloseq/$sample.biom ../12_phyloseq/example.biom
done
```

```{bash, engine.opts='-l'}
# Let's go to our work directory
cd /xdisk/bhurwitz/bh_class/${NETID}/assignments/12_phyloseq
```

```{r}
# check that we are in the correct directory
get_wd()
```

```{r message=FALSE}
# loading all of the libraries we need to view the biom files we created
library("phyloseq")
library("ggplot2")
library("RColorBrewer")
library("patchwork")
```

## Getting started and cleaning up the data

To get started, we need to import our example.biom file into R and work with it using the phyloseq package 

```{r}
# View(merged_annotations)
merged_annotations <- import_biom("example.biom")
merged_annotations
```

To remove unnecessary characters in `.Data` (matrix), we will use the command `substring()`. 
This command helps extract or replace characters in a vector. 

```{r}
merged_annotations@tax_table@.Data <- substring(merged_metagenomes@tax_table@.Data, 4)
```

Next each character in `.Data` is preceded by three spaces occupied by a letter and two underscores, 
for example: `o__Rhodobacterales`. In this case, "Rodobacterales" starts at position 4 with an R. 
So, to remove the unnecessary characters, we will use the following code:

```{r}
colnames(merged_metagenomes@tax_table@.Data)<- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
```

## Using phyloseq objects

Let's take a look at one of the exmaple.biom files in phyloseq to look at differences in annotation for 
each assembly and for the read files.

We will convert this file into a `phyloseq` S4 object. This object contains processed microbiota data from our FMT study,
with taxonomic annotation for the megahit & metaspades contigs, and reads.

What does the phyloseq object have in it?

The printed object shows you functions you can use to access the data inside.

You can also use the `@` symbol (see below).

```{r}
tax_table(merged_annotations) %>% head()
```

```{r}
rank_names(merged_annotations)
```

```{r}
otu_table(merged_annotations)[1:15, 1:8] 
# merged_annotations@otu_table[1:15, 1:10] # the same result
```

```{r}
sample_variables(merged_annotations)
```

```{r}
sample_data(merged_annotations)[1:15, 1:5]
```

```{r}
sample_names(merged_annotations) %>% head(10) 
```

We will use a command named `unique()` to explore how many phyla we have. Let us see the result
we obtain from the following code:

```{r}
> unique(merged_annotations@tax_table@.Data[,"Phylum"])
```

We know that the gut microbiome has common phyla. LetÂ´s use the command `sum()` to ask R to add these up for us:

```{r}
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Firmicutes")
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Bacteroidetes")
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Actinobacteria")
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Lactobacillales")
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Verrucomicrobia")
sum(merged_annotationss@tax_table@.Data[,"Phylum"] == "Proteobacteria")
```

Now let's see how many taxa are in that phylum:

```{r}
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Firmicutes", "Class"])
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Bacteroidetes", "Class"])
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Actinobacteria", "Class"])
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Lactobacillales", "Class"])
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Verrucomicrobia", "Class"])
unique(merged_metagenomes@tax_table@.Data[merged_metagenomes@tax_table@.Data[,"Phylum"] == "Proteobacteria", "Class"])
```

Great job! Now you know more about working with phyloseq objects. We will use these in later lessons to
analyze our FMT datasets.

Copy your notebook in to the current working directory.

```{bash, engine.opts='-l'}
# copy your notebook into the working directory
cp ~/12_phyloseq.Rmd .
```
