<!DOCTYPE html>
<html>
<head>
<title>09_metag_binning.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="metagenome-binning">Metagenome Binning</h1>
<h3 id="questions">Questions:</h3>
<ul>
<li>How can we obtain the original genomes from a metagenome?</li>
</ul>
<h3 id="objectives">Objectives:</h3>
<ul>
<li>Obtain Metagenome-Assembled Genomes from the metagenomic assembly.</li>
</ul>
<h3 id="keypoints">Keypoints:</h3>
<ul>
<li>Metagenome-Assembled Genomes (MAGs) sometimes are obtained from curated contigs grouped into bins.</li>
<li>Use MAXBIN to assign the contigs to bins of different taxa.</li>
</ul>
<h2 id="metagenomic-binning">Metagenomic binning</h2>
<p>Contigs in the assembly can be separated into bins with a process called binning. This process allows for a separate analysis of each species contained in the metagenome with enough reads to reconstruct a genome. Genomes reconstructed from metagenomic assemblies are called MAGs (Metagenome-Assembled Genomes).
In this process, the assembled contigs from the metagenome will be assigned to different bins (FASTA files that contain certain contigs). Ideally, each bin corresponds to only one original genome (a MAG).</p>
<a href="../fig/03-05-01.png">
  <img src="../fig/03-05-01.png" width="435" height="631" alt="Diagram depicts the DNA sequences in the original sample as circular chromosomes of three different taxa. After sequencing, the DNA sequences of the three different taxa are mixed as small linear reads; after the assembly, we have contigs, each corresponding to a single taxon, except for the ones with a bad assembly that has sequences of different taxa in the same contig, after the binning taxa separate the contigs."/>
</a>
<p>An obvious way to separate contigs that correspond to a different species is by their taxonomic assignation, but
there are more reliable methods that do the binning using
characteristics of the contigs, such as their GC content, the use of tetranucleotides (composition), or their coverage (abundance).</p>
<p><a href="https://sourceforge.net/projects/maxbin/files/">Maxbin</a> is a binning algorithm
that distinguishes between contigs that belong to different bins according to their
coverage levels and the tetranucleotide frequencies they have.</p>
<blockquote>
<h2 id="discussion-1-relation-between-mags-and-depth">Discussion 1: Relation between MAGs and depth</h2>
<p>The sequencing center has returned a file to you with 18,412 reads. Given that the bacterial genome size range
between 4Mbp and 13Mbp (Mb=10^6 bp) and that the size of the reads in this run is 150bp. With these data,
how many complete bacterial genomes can you reconstruct?</p>
</blockquote>
<details>
  <summary markdown="span">Solution</summary>
  <ul> 
None, because 18,412 reads of 150bp gives a total count of 2,761,800 bp (~2Mbp). Even if no read maps to the same region, the amount of base pairs is less than the size of a bacterial genome. A “typical” bacterial genome is around 5 million bp.
</details>
<br>
<p>Let's bin the sample we just assembled. The command for running MaxBin is <code>run_MaxBin.pl</code>, and the arguments it needs are the FASTA file of the assembly, the FASTQ with the forward and reverse reads, the output directory, and the name.</p>
<p>/contrib/singularity/shared/bhurwitz/maxbin2:2.2.7--hdbdd923_5.sif run_MaxBin.pl --help</p>
<pre class="hljs"><code><div>$ cd /xdisk/bhurwitz/bh_class/YOUR_NETID/exercises/08_assembly/assembly_JC1A
$ mkdir ../../09_metag_binning/assembly_JC1A
$ /contrib/singularity/shared/bhurwitz/maxbin2:2.2.7--hdbdd923_5.sif run_MaxBin.pl \
-thread 8 -contig JC1A_contigs.fasta \
-reads ../../06_qc_trimming/JC1A_R1.trim.fastq.gz \
-reads2 ../../06_qc_trimming/JC1A_R2.trim.fastq.gz \
-out ../../09_metag_binning/assembly_JC1A
</div></code></pre>
<pre class="hljs"><code><div>MaxBin 2.2.7
Thread: 8
Input contig: JC1A_contigs.fasta
Located reads file [../../06_qc_trimming/JC1A_R1.trim.fastq.gz]
Located reads file [../../06_qc_trimming/JC1A_R2.trim.fastq.gz]
out header: ../../09_metag_binning/assembly_JC1A
Running Bowtie2 on reads file [../../06_qc_trimming/JC1A_R1.trim.fastq.gz]...this may take a while...
Reading SAM file to estimate abundance values...
Running Bowtie2 on reads file [../../06_qc_trimming/JC1A_R2.trim.fastq.gz]...this may take a while...
Reading SAM file to estimate abundance values...
Searching against 107 marker genes to find starting seed contigs for [JC1A_contigs.fasta]...
Running FragGeneScan....
Running HMMER hmmsearch....
Try harder to dig out marker genes from contigs.
Marker gene search reveals that the dataset cannot be binned (the medium of marker gene number &lt;= 1). Program stop.
</div></code></pre>
<p>It seems impossible to bin our assembly because the number of marker genes is less than 1.
We could have expected this as we know it is a small sample.</p>
<p>We will perform the binning process with the other sample from the same study that is larger. We have the assembly precomputed in the <code>/xdisk/bhurwitz/bh_class/mags/</code> directory.</p>
<pre class="hljs"><code><div>$ cd ../../08_assembly
$ mkdir assembly_JP4D
$ cd assembly_JP4D
$ cp /xdisk/bhurwitz/bh_class/mags/JP4D_contigs.fasta .
$ mkdir ../../09_metag_binning/assembly_JP4D
$ /contrib/singularity/shared/bhurwitz/maxbin2:2.2.7--hdbdd923_5.sif run_MaxBin.pl \
-thread 8 -contig JP4D_contigs.fasta \
-reads ../../06_qc_trimming/JP4D_R1.trim.fastq.gz \
-reads2 ../../06_qc_trimming/JP4D_R2.trim.fastq.gz \
-out ../../09_metag_binning/assembly_JP4D
</div></code></pre>
<p>It will take ~7 min minutes to run. Moreover, it will finish with an output like this:</p>
<pre class="hljs"><code><div>========== Job finished ==========
Yielded 4 bins for contig (scaffold) file JP4D_contigs.fasta

Here are the output files for this run.
Please refer to the README file for further details.

Summary file: ../../09_metag_binning/assembly_JP4D.summary
Genome abundance info file: ../../09_metag_binning/assembly_JP4D.abundance
Marker counts: ../../09_metag_binning/assembly_JP4D.marker
Marker genes for each bin: ../../09_metag_binning/assembly_JP4D.marker_of_each_gene.tar.gz
Bin files: ../../09_metag_binning/assembly_JP4D.001.fasta - ../../09_metag_binning/assembly_JP4D.004.fasta
Unbinned sequences: ../../09_metag_binning/assembly_JP4D.noclass

Store abundance information of reads file [../../06_qc_trimming/JP4D_R1.trim.fastq.gz] in [../../09_metag_binning/assembly_JP4D.abund1].
Store abundance information of reads file [../../06_qc_trimming/JP4D_R2.trim.fastq.gz] in [../../09_metag_binning/assembly_JP4D.abund2].

========== Elapsed Time ==========
0 hours 6 minutes and 18 seconds.

</div></code></pre>
<p>With the <code>.summary</code> file, we can quickly look at the bins that MaxBin produced.</p>
<pre class="hljs"><code><div>$ cat ../../09_metag_binning/assembly_JP4D.summary
</div></code></pre>
<pre class="hljs"><code><div>Bin name	Completeness	Genome size	GC content
assembly_JP4D.001.fasta	57.9%	3141556	55.5
assembly_JP4D.002.fasta	87.9%	6186438	67.3
assembly_JP4D.003.fasta	51.4%	3289972	48.1
assembly_JP4D.004.fasta	77.6%	5692657	38.9
</div></code></pre>

</body>
</html>
