---
title: "Microbiome data analysis with microViz"
subtitle: "Working with barcharts"
author: "David Barnett"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
```

```{r message=FALSE}
library(seriation)
library(dplyr)
library(purrr)
library(ggplot2)
library(phyloseq)
library(microViz)
library(shiny)
```

```{r}
mice <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/mice.rds')
```

Now that we are getting into microbiome analytics, we are going to transition to using more complete datasets with taxonomic data and metadata.
We'll start by using a gut microbiome from a study of antibiotic administration in mice.

-   `mice` is 16S rRNA gene amplicon sequencing data, from the mouse antibiotics study
    -   They used Illumina MiSeq and processed the data into ASVs using DADA2

The data has already been processed from fastq files into counts per taxon, and are available to you on the UA cluster.
This is the exciting bit, you get to explore and visualize the data, and do statistics (yay!).

# Intro to phyloseq

This is a `phyloseq` S4 object, containing processed microbiota data from the mouse study.

<details>

<summary>

**Click here for more details about the mice dataset:**

</summary>

**The mice data**

The data originate from a study on the effects of oral antibiotic administration on flavivirus infection (<https://www.ncbi.nlm.nih.gov/pubmed/29590614>).
Sequence data was generated from extracted nucleic acid from stool samples collected from individually caged mice and amplified using primers specific for the V4 region using primers 515F/806R.

The study followed flavivirus infection after the following treatments:

1)  Koolaid: Antibiotics are provided to the mice via their drinking water. As many of the antibiotics taste bad, koolaid is added as a sweetener. Therefore, the appropriate control is water spiked and labelled *koolaid*.
2)  Ampicillin (Amp): <https://en.wikipedia.org/wiki/Ampicillin>
3)  Metronidazole (Met): <https://en.wikipedia.org/wiki/Metronidazole>
4)  Ampicillin + Metronidazole (Amp+Metro)

Treatments were supplied ad libitum for 2 weeks prior to viral infection and maintained for 2 weeks post-infection.
Primary outcome was mouse survival.
Each treatment group had two subgroups of mice that were either a) left uninfected as controls or b) infected with West Nile Virus via a subcutaneous foot pad injection.

</details>

Get a little familiar with the object.
What does it have in it?
Can you look at each part?

The printed object shows you functions you can use to access the data inside.

You can also use the `@` symbol.

```{r}
mice
# View(mice)
```

```{r}
tax_table(mice) %>% head()
```

```{r}
rank_names(mice)
```

```{r}
otu_table(mice)[1:15, 1:8] 
# mice@otu_table[1:15, 1:10] # the same result
```

```{r}
sample_variables(mice)
```

```{r}
sample_data(mice)[1:15, 1:5]
```

```{r}
sample_names(mice) %>% head(10) 
```

# Looking at microbiome data

Okay, so how do we look at the microbiota abundance data?
To do this, we're going to use the R package `microViz`

## Barcharts: bad to better

Lets take a very small subset of this data to get started.
Just the control group (vehicle treatment) at day 13.

```{r}
# We can filter the samples like this, using the sample_data information
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle')
```


### Bad bars

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
  comp_barplot(
    tax_level = 'unique', n_taxa = 12, bar_width = 0.7,
    sample_order = 'asis', tax_transform_for_plot = 'identity'
  ) +
  coord_flip() 
```

What is going on here??
The unique taxa have uninformative IDs, and we also got a message about problems with the taxonomy table.

The total number of reads also varies a lot between samples!
The total number of reads for each sample is NOT a reliable indicator of the biomass or bacterial load of each sample.
So for now we will just consider the relative abundance of each taxon, as proportions of the total counts for that sample.

### Compositions (%)

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
  comp_barplot(
    tax_level = 'unique', n_taxa = 12, sample_order = 'asis', bar_width = 0.7
  ) +
  coord_flip()
```

### Fixing tax_table 

Let's look at the taxonomy table interactively

```{r}
# tax_fix_interactive(mice) # run this in the R Console for an interactive look
```

Looks like we just need to fill in some blank cells when a sequence was not classified at genus or family.
`tax_fix` can do this, it just copies down info from a higher rank classification.
Let's update our `mice` phyloseq object with this fix.

```{r}
mice <- tax_fix(mice, verbose = FALSE)
```

We can also rename the unique taxa with a more informative name, according to their classification at the rank of Family (and how common they are).

```{r}
mice %>% taxa_names() %>% head
mice <- tax_rename(mice, rank = 'Family')
mice %>% taxa_names() %>% head
```

Let's try again with the better names.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
  comp_barplot(
    tax_level = 'unique', n_taxa = 12, sample_order = 'asis', bar_width = 0.7
    ) +
  coord_flip()
```

Sadly we don't have enough distinct colours to show all the unique taxa.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
  comp_barplot(
    tax_level = 'unique', n_taxa = 12, sample_order = 'asis', bar_width = 0.7,
    merge_other = FALSE
    ) +
  coord_flip()
```

### Aggregating taxa

So let's "aggregate" all the counts into family-level groups.
For each family, this adds together all the counts from every ASV that belongs to that family. We can do that by changing the `tax_level` argument to "Family".

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
  comp_barplot(
    tax_level = "Family", n_taxa = 10, bar_width = 0.7, sample_order = 'asis'
  ) +
  coord_flip()
```

By aggregating at family level, we have sacrificed taxonomic resolution, compared to using ASVs.
But this way we can get an idea of which families are the most abundant, and how variable the communities are.

Try making some similar plots aggregated at different taxonomic ranks.

```{r}
# rank_names(mice)
# mice %>%
#   ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>%
#   comp_barplot(tax_level = , n_taxa = 10, sample_order = 'asis', merge_other = FALSE)
```

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>% 
  comp_barplot(
    tax_level = "Genus", n_taxa = 12, bar_width = 0.7,
    sample_order = 'asis', merge_other = FALSE
  ) +
  coord_flip()
```

Many of the ASVs in this mice data, the *Porphyromonadaceae*, could not be classified at genus level.

```{r}
mice %>% 
  ps_filter(treatment_days == 'D13', virus == 'WNV2000', treatment == 'Vehicle') %>% 
  comp_barplot(
    tax_level = "Phylum", n_taxa = 7, bar_width = 0.7, sample_order = 'asis'
  ) +
  coord_flip()
```


### Fickle Phyla

A note on phylum names!
There have been major changes this year and some of these are now old names.
Most published research is of course with the old names (and still probably will be for a year or so).

# Fun with barcharts

More examples/tutorial of visualizing microbiome data using stacked barcharts can be found here:
https://david-barnett.github.io/microViz/articles/web-only/compositions.html

Try it out for yourself a bit!

Bar charts often look better when you sort the samples by similarity.
The webpage mentions using Bray-Curtis distances and hierarchical clustering to 
sort samples. We haven't discussed dissimilarity or distances yet, but we will in the next session today!

For now, just appreciate that it can make the bar charts easier to read!

```{r}

```

# Session info

Records your package versions etc.
Useful for debugging / reproducing analysis.

```{r}
devtools::session_info()
```
