---
title: "Microbiome data analysis with microViz"
subtitle: "Dissimilarity, ordination, & filtering"
author: "David Barnett"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: true
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, dpi = 120)
knitr::opts_knit$set(root.dir = '/xdisk/bhurwitz/bh_class/NETID/exercises/17_ordination') # Change your NETID
```

```{r message=FALSE}
library(seriation)
library(dplyr)
library(purrr)
library(ggplot2)
library(phyloseq)
library(microViz)
library(shiny)
```

```{r}
shao19 <- readRDS('/xdisk/bhurwitz/bh_class/data/microviz/shao19.rds')
```

```{r}
shao4 <- shao19 %>% ps_filter(family_role == "child", infant_age == 4)
```

## Ordination

What can we do with these distances or dissimilarities?
We can make an ordination.

Ordination refers to the process of ordering things (in our case: samples), so that similar things (samples) are closer to each other, 
and dissimilar things (samples) are further away.

### PCoA

Principal Co-ordinates Analysis is one kind of ordination.

Takes distance matrix and finds new dimensions (a co-ordinate system if you like).
The new dimensions are created with the aim to preserve the original distances between samples, 
but to capture the majority of this distance information in the first few dimensions.
We like this because it makes it easier to visualize the patterns in your data (in 2D or 3D).

There is helpful info about ordination methods including PCoA on
the GUide to STatistical Analysis in Microbial Ecology (GUSTA ME) website.
<https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis>
This website covers a lot of topics, we'll cover PCoA now, and PCA in the next session.
The other topics may be interesting for you to read at a later date if you'll work on microbiome analysis.

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_agg(rank = "genus") %>%
  dist_calc(dist = "bray") %>%
  ord_calc(method = "PCoA") %>%
  ord_plot(alpha = 0.6, size = 2) +
  theme_classic(12) +
  coord_fixed(0.7)
```

To get a little insight into what has happened here, we can colour each sample according to its dominant (most abundant) genus.

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  ps_calc_dominant(rank = "genus", none = "Mixed", other = "Other") %>%
  tax_agg(rank = "genus") %>%
  dist_calc(dist = "bray") %>%
  ord_calc(method = "PCoA") %>%
  ord_plot(color = "dominant_genus", alpha = 0.6, size = 2) +
  scale_color_brewer(name = "Dominant Genus", palette = "Dark2") +
  theme_classic(12) +
  coord_fixed(0.7)
```


### Interactive ordination!

`microViz` provides a Shiny app `ord_explore` to interactively create and explore PCoA plots and other ordinations.
See the code below to get started. But read the instructions first.

**Here are a few things to try:**

-   Colour the samples using the variables in the sample data
-   Select a few samples to view their composition on barplots!
-   Change some ordination options:
    -   Different rank of taxonomic aggregation
    -   Different distances we've discussed
-   Copy the automatically generated code
    -   Exit the app (press escape or click red button in R console!)
    -   Paste and run the code to recreate the ordination plot
    -   Customise the plot: change colour scheme, title, etc.
-   Launch the app again with a different subset of the data
    -   Practice using `ps_filter` etc.
    -   e.g. plot the data of the mothers' gut microbiomes!
    -   compute one or more alpha diversity measures


**Beware:**

-   UniFrac distances can be quite slow (over a minute) to calculate!
    -   Filter to fewer samples and fewer taxa to speed it up (Before launching the app)
-   There are many distances available, feel free to try out ones we haven't talked about
    -   BUT:
        -   You shouldn't use a distance that you don't understand in your actual work, even if the plot looks nice! ;)
        -   Some of them might not work...
            -   They are mostly implemented in the package `vegan` and I haven't tested them all
            -   Errors will appear in the RStudio R console
            -   You can report to me any distances that don't work if you're feeling helpful!
-   There are other ordination methods available in `ord_explore`, which we haven't discussed
    -   We will discuss PCA and various transformations a little later 
    -   Some things we won't have time to cover, but you can look here for info on topics like constrained ordination --\> TODO: insert gusta me ecology website link

``` r
# fire up the shiny app
# run these lines in your console (don't keep in script/notebook)
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  # calculate new sample variables with dominant taxon (optional)
  ps_calc_dominant(rank = "genus", none = "Mixed", other = "Other") %>%
  # launch a Shiny app in your web browser!
  ord_explore()
```

``` r
# different options
# run this line in your console
shao19 %>%
  ps_filter(family_role == "mother") %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  # calculate a few sample variables for interest (optional)
  ps_calc_dominant(rank = "genus", none = "Mixed", other = "Other") %>%
  ps_calc_diversity(rank = "genus", index = "shannon") %>%
  ps_calc_richness(rank = "genus", index = "observed") %>%
  # launch a Shiny app in your web browser!
  ord_explore()
```

## PERMANOVA:

Permutational multivariate analysis of variance.

-   ANOVA - analysis of variance (statistical modelling approach)
-   Multivariate - more than one dependent variable (multiple taxa!)
-   Permutational - statistical significance estimates obtained by shuffling the data many times

For more details. See this excellent book chapter by Marti Anderson on PERMANOVA: <https://onlinelibrary.wiley.com/doi/full/10.1002/9781118445112.stat07841>

Sometimes also called NP-MANOVA (non-parametric MANOVA) e.g. on the
GUide to STatistical Analysis in Microbial Ecology [website.](https://sites.google.com/site/mb3gustame/hypothesis-tests/manova/npmanova)

**TLDR:** Are those groups on the PCoA actually different??

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_agg(rank = "genus") %>%
  dist_calc(dist = "bray") %>%
  ord_calc(method = "PCoA") %>%
  ord_plot(alpha = 0.6, size = 2, color = "birth_mode") +
  theme_classic(12) +
  coord_fixed(0.7) +
  stat_ellipse(aes(color = birth_mode)) +
  scale_color_brewer(palette = "Set1")
```

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_agg(rank = "genus") %>%
  dist_calc(dist = "bray") %>%
  dist_permanova(variables = "birth_mode", n_perms = 99, seed = 123) %>%
  perm_get()
# Use more permutations for a more reliable p.value in your real work (slower)
# Set a random seed number for reproducibility of this stochastic method
```

You can see from the model output that the p value, Pr(\>F) is below 0.05.
So there is good statistical evidence that the bacterial gut microbiota composition of c-section delivered infants 
has a different composition than vaginally delivered infants at 4 days of age.

You should also report that you used Bray-Curtis dissimilarities, calculated on genera.
(after keeping only unique taxa with a prevalence of at least 2.5%!)

It's probably a good idea to decide on a couple of appropriate distance measures up front for these tests, 
and report both (at least in supplementary material), as the choice of distance measure can affect results and conclusions!

You can also adjust for covariates in PERMANOVA, and often should, depending on your study design.
Let's fit a more complex model, adjusting for infant sex, birth weight, and the total number of assigned reads.

```{r}
shao4 %>%
  tax_filter(min_prevalence = 2.5 / 100, verbose = FALSE) %>%
  tax_agg(rank = "genus") %>%
  dist_calc(dist = "bray") %>%
  dist_permanova(
    variables = c("birth_mode", "sex", "birth_weight", "number_reads"),
    n_perms = 99, seed = 111
  ) %>%
  perm_get()
# Use more permutations for a more reliable p.value in your real work (slower)
# Set a random seed number for reproducibility of this stochastic method
```


# Session info

Records your package versions etc.
Useful for debugging / reproducing analysis.

```{r}
devtools::session_info()
```
